### PR TITLE
feat: injector and injections unit. tests and sample WASM file

### PR DESCRIPTION
**Pull request by mikirov**
- Separate FunctionMapper functionality into separate getters so they can be tested
- implement unit tests based on sample WASM file
- implement unit tests for correct instructions being present in target WASM after injecting

### CODE REVIEW

File: src/injecting/injections.rs [hunk 1]
```diff
@@ -107,3 +107,110 @@ fn inject_heap_overflow(module: &mut Module) -> Result<(), String> {
...
+    #[test]
+    fn test_inject_heap_overflow() {
+        let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        d.push("test.wasm");
+        let mut module = load_module_from_wasm(d.as_path()).unwrap();
+
+        let injection = Injection::HeapOverflow;
+        assert!(injection.inject(&mut module).is_ok());
+
+        let index: u32 = module.get_malloc_index().unwrap() as u32;
```
Comment (reo101) [inline src/injecting/injections.rs:206]: Unnecessary type annotation

File: src/injecting/injector.rs [hunk 2]
```diff
@@ -72,20 +93,92 @@ impl FunctionMapper for Module {
...
+    // NOTE VALUES ARE FOR TEST WASM ONLY AND WILL DIFFER FOR DIFFERENT WASM BLOBS!!!
+    const VALIDATE_BLOCK_GLOBAL_INDEX: usize = 1733;
+    const IMPORT_SECTION_LENGTH: usize = 39;
+    const MALLOC_INDEX: usize = 25;
+    const WASM_INSTRUCTION_COUNT: usize = 358;
+
+    #[test]
+    fn test_get_global_function_index() {
+        let mut d = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        d.push("test.wasm");
```
Comment (reo101) [inline src/injecting/injector.rs:146]: Might want to extract this `PathBuf` (or even the loaded module for that matter) as a constant. You can test the module loading in isolation.

File: test.wasm
Comment (reo101) [inline test.wasm:1]: Generally not a good idea to store binary blobs in git (consider using git-lfs or similar)

File: src/injecting/injector.rs [hunk 3]
```diff
@@ -72,20 +93,92 @@ impl FunctionMapper for Module {
...
-            .code_section_mut()
-            .ok_or("No code section")?
-            .bodies_mut()
-            .get_mut(local_function_index)
-            .ok_or(format!(
-                "Function '{}' not found in the code section",
-                function_name
-            ))?;
+            .get_function_body(local_function_index, function_name)
+            .unwrap();
```
Comment (reo101) [inline src/injecting/injector.rs:122]: You're are already in a function returning a `Result`. Please propagate the errors correctly (no `unwrap`s). Applies to the above 4 statements in this function.

File: src/injecting/injections.rs [hunk 4]
```diff
@@ -107,3 +107,110 @@ fn inject_heap_overflow(module: &mut Module) -> Result<(), String> {
...
+        let function_body = get_function_body(&mut module);
+
+        let expected = vec![[Instruction::I32Const(33_554_431), Instruction::Call(index)]; 8]
+            .into_iter()
+            .flatten()
+            .collect::<Vec<Instruction>>();
```
Comment (reo101) [inline src/injecting/injections.rs:212]: Clever monadic usage, nice :smile:

You could also consider something like this, for readability:

```rust
let expected = vec![Instruction::I32Const(33_554_431), Instruction::Call(index)]
  .into_iter()
  .cycle()
  .take(16)
  .collect::<Vec<_>>();
```

File: src/injecting/injector.rs [hunk 5]
```diff
@@ -72,20 +93,92 @@ impl FunctionMapper for Module {
...
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::util::load_module_from_wasm;
+    use std::path::PathBuf;
+
+    // NOTE VALUES ARE FOR TEST WASM ONLY AND WILL DIFFER FOR DIFFERENT WASM BLOBS!!!
```
Comment (reo101) [inline src/injecting/injector.rs:137]: You could use `// NOTE: ...` to be consistent with the rest of the comments in the project. Also, consider using a higher severity - like `WARNING`.

File: src/injecting/injector.rs [hunk 6]
```diff
@@ -72,20 +93,92 @@ impl FunctionMapper for Module {
...
+
+        // Count the number of imported functions listed by the module
+        let import_section_len: usize = self.get_import_section_len().unwrap();
+
+        let malloc_index = self.get_malloc_index().unwrap();
+
+        // Compute the local function index (for the code section) by subtracting
+        // the number of imported functions from the global function index
+        let local_function_index = global_function_index - import_section_len;
         // Extract the `${function_name}` function_body
```
Comment (reo101) [inline src/injecting/injector.rs:119]: A newline for consistency would be nice here :smile:

File: src/injecting/injector.rs [hunk 7]
```diff
@@ -72,20 +93,92 @@ impl FunctionMapper for Module {
...
+        // functions. So we actually need the imported function count to resolve the
+        // actual index of the given function in own functions list.
+
+        // Find the global function index
+        let global_function_index = self.get_global_function_index(function_name).unwrap();
+
+        // Count the number of imported functions listed by the module
+        let import_section_len: usize = self.get_import_section_len().unwrap();
+
+        let malloc_index = self.get_malloc_index().unwrap();
```
Comment (reo101) [inline src/injecting/injector.rs:114]: You could either add a short comment here (regarding why this is complex enough that it needs to be abstracted away in a function).

I've commented about the use of `.unwrap()` below.

**Review by reo101 — CHANGES_REQUESTED** [review]
Good job on the PR :smile:. I've left a few requests for changes

File: test.wasm
Comment (mikirov) [reply test.wasm:1]: True, however for one file i was willing to make an exception. If we go to git-lfs there is no easy way to go back (at least in by experience it was clunky). Do we need it for something other than this file? Potentially in the future?

File: src/injecting/injections.rs [hunk 8]
```diff
@@ -196,20 +196,19 @@ mod tests {
...
 
-        let expected = vec![[Instruction::I32Const(33_554_431), Instruction::Call(index)]; 8]
-            .into_iter()
-            .flatten()
-            .collect::<Vec<Instruction>>();
+        let expected = vec![Instruction::I32Const(33_554_431), Instruction::Call(index)]
+        .into_iter()
+        .cycle()
+        .take(16)
+        .collect::<Vec<_>>();
```
Comment (reo101) [inline src/injecting/injections.rs:211]: Please format this

File: src/injecting/injector.rs [hunk 9]
```diff
@@ -106,20 +106,32 @@ impl FunctionMapper for Module {
...
+            Ok(import_section_len) => import_section_len,
+            Err(err) => return Err(err),
+        };
 
-        let malloc_index = self.get_malloc_index().unwrap();
+        // get the global function index of the dynamic memory allocation method
+        let malloc_index = match self.get_malloc_index() {
+            Ok(malloc_index) => malloc_index,
+            Err(err) => return Err(err),
+        };
```
Comment (reo101) [inline src/injecting/injector.rs]: This is redundant, you can use the `?` shorthand:

```rust
// Get the global function index of the dynamic memory allocation method
let malloc_index = self.get_malloc_index()?;
```

(Applies to the whole function)

**Review by reo101 — CHANGES_REQUESTED** [review]

File: test.wasm
Comment (reo101) [reply test.wasm:1]: Fair, if it is only one fine - yes, no problem. In the future we might get around to testing lots of different wasms which would be supplied in other ways (maybe a directory filled with fresh ones, ready for injection and testing).

But alas I digress, it's fine for now / can be fixed up later down the line (would still remain in git history tho).

Comment (dimitrovmaksim) [reply test.wasm:1]: I'll also suggest moving it into a folder (for example `wasm`) and not leaving the file in the root

**Review by reo101 — CHANGES_REQUESTED** [review]
Apart from that last comment, LGTM

File: src/injecting/injector.rs [hunk 10]
```diff
@@ -131,38 +131,38 @@ impl FunctionMapper for Module {
...
     use crate::util::load_module_from_wasm;
-    use std::path::PathBuf;
+    use std::path::Path;
 
     /// WARNING: VALUES ARE FOR TEST WASM ONLY AND WILL DIFFER FOR DIFFERENT WASM BLOBS!!!
     const VALIDATE_BLOCK_GLOBAL_INDEX: usize = 1733;
     const IMPORT_SECTION_LENGTH: usize = 39;
     const MALLOC_INDEX: usize = 25;
     const WASM_INSTRUCTION_COUNT: usize = 358;
+    const WASM_PATH: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/wasm/test.wasm");
```
Comment (dimitrovmaksim) [inline src/injecting/injector.rs:141]: The `wasm` folder is ingored in the .gitignore file. Now the file is no present in the repo.

**Review by reo101 — APPROVED** [review]
LGTM

**Comment by reo101** [comment]
There's a typo in the commit name

**Pull request merged**